aiida_quantumespresso/utils/restart.py

```diff
@@ -1,451 +1,57 @@
 # -*- coding: utf-8 -*-
-"""
-Utility functions to return builders ready to be submitted
-for restarting a Quantum ESPRESSO calculation (or to apply small
-modifications).
-"""
+"""Utility functions to return process builders ready to be submitted for restarting a Quantum ESPRESSO calculation."""
 from __future__ import absolute_import
-from aiida.common import InputValidationError
-from aiida.common.links import LinkType
 
+from aiida_quantumespresso.calculations.cp import CpCalculation
+from aiida_quantumespresso.calculations.neb import NebCalculation
+from aiida_quantumespresso.calculations.ph import PhCalculation
+from aiida_quantumespresso.calculations.pw import PwCalculation
 
-def clone_calculation(calculation):
-    """
-    Create a clone of a Calculation node.
-    This is a temporary workaround, it's better to create a new builder.
-
-    :returns: an unstored clone of this node. Note: links are not copied
-    """
-    clone = self.__class__()
-    clone.dbnode.dbcomputer = self._dbnode.dbcomputer
-    clone.dbnode.type = self._dbnode.type
-    clone.label = self.label
-    clone.description = self.description
-
-    for key, value in self.iterattrs():
-        if key != Sealable.SEALED_KEY:
-            clone.set_attribute(key, value)
-
-    for path in self.get_folder_list():
-        clone.add_path(self.get_abs_path(path), path)
-
-    return clone
-
-
-def _create_restart_pw_cp(
-    parent_calc, force_restart, parent_folder_symlink, use_output_structure, restart_from_beginning
-):
-    """
-    Function to restart a pw.x or cp.x calculation that was not completed before
-    (like max walltime reached...) i.e. not to restart a really FAILED calculation.
-
-    It returns a `builder` for a new calculation, with all links prepared
-    but not stored in DB.
-    To submit it, simply call::
-
-        from aiida.engine import submit
-        submit(builder)
-
-    :param parent_calc: the calculation you want to restart
-    :param bool force_restart: restart also if parent is not in FINISHED
-       state (e.g. FAILED, IMPORTED, etc.).
-    :param bool parent_folder_symlink: if True, symlinks are used
-       instead of hard copies of the files.
-       Pass None for the default calculation behavior.
-    :param bool use_output_structure: if True, the output structure
-       of the restarted calculation is used if available, rather than its
-       input structure. Useful for nscf or bands calculations, but it
-       SHOULD NOT be used for the restart of a relaxation run.
```

Comment:
Indeed the new code does not use the output structure of the previous calculation. But is this not appropriate to use it?

Reply:
I think that adding an option to do this is not worth it, because it would take the user literally one line to use it:
```
calculation = load_node()
builder = get_builder_restart(calculation)
builder.structure = calculation.outputs.output_structure
```
Adding a bunch of logic to deal with calculations that then might not have one and substituting with the input structure (which is done by default through the `node.get_builder_restart` call anyway) seems pointless

Reply:
Ok, but I was wondering about this part of the docstring: `[...] it SHOULD NOT be used for the restart of a relaxation run`.
Does it mean that you should never use the output structure of a relax calculation to restart it? If so, why?

Reply:
Yeah it doesn't make sense. I guess what they were trying to say is that if you have a completed "relax" `PwCalculation` and then want to restart from it to do an "nscf" or "bands, then you want to use the output structure of course, which is what the function argument is for. However, even for a "relax" calculation, it might still make sense to reuse the output structure, e.g. if it did not converge yet, but already performed some relaxation steps. Another reason why I dropped the `use_output_structure` argument

Reply:
Ah I see.
Ok, thanks (also for the offline discussion).